# Advent of Code - Day 2

In [1]:
import os
import json

## Load File Content First

In [2]:

FILE_PATH = "input.txt"

def open_file(file_path: str) -> list:
    """
    Open file
    The file is a txt file
    """
    with open(file_path, 'r') as file:
        content = [line.strip() for line in file]
    return content

FILE_CONTENT = open_file(FILE_PATH)
assert isinstance(FILE_CONTENT, list)
content_size = len(FILE_CONTENT)
print(f"Length of file content: {content_size}")

Length of file content: 1000


## Strategy

1. Load the data
2. Iterate through each line from the data
3. Might need to split the data and convert in a list
4. Check if the list is Asc or Desc sorted
5. If Yes in 4, then move to the interval calculation, else ignore
6. Calculate "safe"

In [3]:
# Analyze the data
t = FILE_CONTENT[0]
print(t)
print(t.split())

# Convert to int
t = [int(x) for x in t.split()]
print(t)

# Make this as a helper function
def convert_to_int(data: str) -> list:
    return [int(x) for x in data.split()]


65 68 71 72 71
['65', '68', '71', '72', '71']
[65, 68, 71, 72, 71]


In [4]:
# Iterate through the file content
map_data = {}
for idx, line in enumerate(FILE_CONTENT):
    list_data = convert_to_int(line)
    map_data[idx] = list_data

assert len(map_data) == content_size

In [5]:
# Write a helper function to check sorted state
def is_sorted(lst: list) -> bool:
    return lst == sorted(lst) or lst == sorted(lst, reverse=True)

In [6]:
# Iterate through the map data, and check if the list is sorted
# If it is, then put it in a new list called sorted_lists

sorted_lists = []
for key, value in map_data.items():
    assert isinstance(value, list)
    if is_sorted(value):
        sorted_lists.append(value)

print(f"Total sorted lists: {len(sorted_lists)}")

Total sorted lists: 640


In [7]:
# Okays, now we have the sorted lists
# We need the adjacent difference calculation for each list

# For EACH list, calculate the difference between adjacent elements
# If the difference is minimum 1 and maximum 3, then it is a valid list
# Else, break the loop and move to the next list

# Note that the lists are either ascending or descending sorted, so the difference should be absolute

# All Vald lists are stored in valid_lists

def get_valid_lists(sorted_lists: list) -> list:
    valid_lists = []
    for lst in sorted_lists:
        valid = True
        for idx in range(1, len(lst)):
            diff = abs(lst[idx] - lst[idx-1])
            if diff < 1 or diff > 3:
                valid = False
                break
        if valid:
            valid_lists.append(lst)

    return valid_lists

In [8]:
valid_lists = get_valid_lists(sorted_lists)

In [9]:
print(f"Total valid lists: {len(valid_lists)}")

Total valid lists: 269


## Part 1 is correct.

### Total Safe Lists are 269 !!

## Part 2

### Tolerate 1 difference error

In [10]:
# okay so part 2 is complex, we will probably brute force it for now, and then if time permits, try to optimize it.append

In [11]:
# basically, here's what we need to do:
# Context:
# - We have the is_sorted function, which checks if a list is sorted
# - We have the get_valid_lists function, which checks if a list is valid.

# Let's modify the get_valid_lists function to return a boolean value if the list is valid or not, and it will only take a list as input

# Before that, we need to start iterating EACH list, remove 1 element at a time, and check if the list is valid or not.


# Function structure:

# 1. Iterate through all the lists
# 2. If the list is sorted, then check if the list is SAFE (valid).
# 3. If the list is not SAFE, then remove 1 element at a time, and check if the list is SAFE or sorted.

# If the list is SAFE, then add it to the valid lists

def check_safe_report_unit(lst: list) -> bool:
    """
    Check if the list is safe or not
    """
    safe = True

    if not is_sorted(lst):
        return False
    
    for idx in range(1, len(lst)):
        diff = abs(lst[idx] - lst[idx-1])
        if diff < 1 or diff > 3:
            safe = False
            break
    
    return safe

def check_safe_report(lst: list) -> bool:
    """
    Remove one element at a time, and check if the list is safe
    """
    for idx in range(len(lst)):
        temp_lst = lst.copy()
        temp_lst.pop(idx)
        if check_safe_report_unit(temp_lst):
            return True
    return False

def get_valid_lists_v2(map_data: dict) -> int:
    safe_v2_lists = []
    for _, lst in map_data.items():
        if check_safe_report(lst):
            safe_v2_lists.append(lst)
    
    return len(safe_v2_lists)

get_valid_lists_v2(map_data)

337

In [12]:
print(f"Total Safe Lists: {get_valid_lists_v2(map_data)}")

Total Safe Lists: 337


## Part 2 is correct!